internal/loader/configuration_client_manager.go

```diff
@@ -0,0 +1,416 @@
+	MinBackoffDuration                  time.Duration = time.Second * 30
+	JitterRatio                         float64       = 0.25
+	SafeShiftLimit                      int           = 63
+)
+
+var (
+	clientOptionWithModuleInfo *azappconfig.ClientOptions = &azappconfig.ClientOptions{
+		ClientOptions: policy.ClientOptions{
+			Telemetry: policy.TelemetryOptions{
+				ApplicationID: fmt.Sprintf("%s/%s", properties.ModuleName, properties.ModuleVersion),
+			},
+		},
+	}
+)
+
+func NewConfigurationClientManager(ctx context.Context, provider acpv1.AzureAppConfigurationProvider) (ClientManager, error) {
+	manager := &ConfigurationClientManager{
+		ReplicaDiscoveryEnabled: provider.Spec.ReplicaDiscovery,
+	}
+
+	var err error
+	var staticClient *azappconfig.Client
+	if provider.Spec.ConnectionStringReference != nil {
+		connectionString, err := getConnectionStringParameter(ctx, types.NamespacedName{Namespace: provider.Namespace, Name: *provider.Spec.ConnectionStringReference})
+		if err != nil {
+			return nil, err
+		}
+		if manager.Endpoint, err = parseConnectionString(connectionString, EndpointSection); err != nil {
+			return nil, err
+		}
+		if manager.secret, err = parseConnectionString(connectionString, SecretSection); err != nil {
+			return nil, err
+		}
+		if manager.id, err = parseConnectionString(connectionString, IdSection); err != nil {
+			return nil, err
+		}
+		if staticClient, err = azappconfig.NewClientFromConnectionString(connectionString, clientOptionWithModuleInfo); err != nil {
+			return nil, err
+		}
+	} else {
+		if manager.Credential, err = CreateTokenCredential(ctx, provider.Spec.Auth, provider.Namespace); err != nil {
+			return nil, err
+		}
+		if staticClient, err = azappconfig.NewClient(*provider.Spec.Endpoint, manager.Credential, clientOptionWithModuleInfo); err != nil {
+			return nil, err
+		}
+		manager.Endpoint = *provider.Spec.Endpoint
+	}
+
+	manager.ValidDomain = getValidDomain(manager.Endpoint)
+	manager.StaticClientWrappers = []*ConfigurationClientWrapper{{
+		Endpoint:       manager.Endpoint,
+		Client:         staticClient,
+		BackOffEndTime: metav1.Time{},
+		FailedAttempts: 0,
+	}}
+
+	return manager, nil
+}
+
+func (manager *ConfigurationClientManager) GetClients() []*ConfigurationClientWrapper {
+	currentTime := metav1.Now()
+	clients := make([]*ConfigurationClientWrapper, 0)
+	for _, clientWrapper := range manager.StaticClientWrappers {
+		if currentTime.After(clientWrapper.BackOffEndTime.Time) {
+			clients = append(clients, clientWrapper)
+		}
+	}
+
+	if !manager.ReplicaDiscoveryEnabled {
+		return clients
+	}
+
+	if currentTime.After(manager.lastFallbackClientAttempt.Time.Add(MinimalClientRefreshInterval)) &&
+		(manager.DynamicClientwrappers == nil ||
+			currentTime.After(manager.lastFallbackClientRefresh.Time.Add(FallbackClientRefreshExpireInterval))) {
+		manager.lastFallbackClientAttempt = currentTime
+		url, _ := url.Parse(manager.Endpoint)
```

Comment:
Seems we didn't validate the Endpoint in Connection String, so this line may return error.